[CK] Extend XDL kernel to support RDNA3/4 (option 2)

linqun opened the pull request:
## Proposed changes

Problem: 
CK contains a lot of well-designed XDL kernel, it is limited on MI cards, CK also have some WMMA kernel for RDNA3/RDNA4, but the supporting is much less XDL kernel.
Consider RDNA3/RDNA4 are more and more used in ML, and our future HW architecture will close to RDNA too, we need improve the support of WMMA kernel.
After compare the WMMA kernel with XDL kernel, it looks the architecture and algorithm is almost same between them. The major differences are:
- XDL kernel based MfmaSelector + Mfma instruction, which doesn't support any WMMA instructions
- XDL kernel assume warp size is constant and it fixed to 64.
  * Template parameter BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave and NXdlPerWave causes warp size is hard code to 64
  * WaveSize is widely used as static constexpr variable, and hardcode to 64.

Proposal:
1. To reduce the efforts of code maintain, I think it is better to extend exist XDL kernel to support both XDL and WMMA, instead of duplicate XDL kernel to WMMA kernel.
2. To eliminate the difference between WMMA and XDL (MFMA), I think we can
- Extend  MfmaSelector + Mfma instruction to support all WMMA instruction.
- Ignore NXdlPerWave in device class template, and calculate it with constexpr according to BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave according to current GPU target.
  * To keep interface compatible, we can continue to keep NXdlPerWave as template parameters.
  * In CK kernel, BlockSize is also used in threadgroup copy parameter, ignore BlockSize requires mode code change.

Solution:
1. Extend  MfmaSelector + Mfma instruction to support all WMMA instruction
- No obvious bottleneck in it, please check the changes in xdlops_gemm.hpp for detail.
- In GFX11, the input data is duplicate in each 16 threads, we need adjust the thread origin data index for input A and B a little.
2. Ignore NXdlPerWave in device class template
- Calculate NXdlPerWave32 and NXdlPerWave64 with according to BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave
- Define GridwiseGemm32 and GridwisGemm64 with NXdlPerWave32 and NXdlPerWave64
- Select GridwiseGemm32/64 according to current GPU target in Invoker::Run
3. Code change in grid/block class
- Calculate WaveSize according to BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave, NXdlPerWave, instead of using get_warp_size
- Add constexpr function GridWiseGemm::IsValidCompilationParameter to check whether the WaveSize in Grid/Block match current HW architecture.
    * It is to avoid generate real kernel with incorrect wavesize, since we defined two kernel in device level.
    * This function can also be used to check other template parameters and avoid unncessary kernel compile, for example
        MPerXDL=32 and NPerXDL = 32 in RDNA3/RDNA4.
4. Split get_warp_size in host and device, in device side, it is constexpr function, in host side, warp size is get from hip driver, and it is a normal function.

Other attempt to Ignore NXdlPerWave
- https://github.com/ROCm/composable_kernel/tree/qlin/merge_wmma_mfma1
 * Remove NXdlPerWave in Grid/Block interface and use function get_warp_size + "BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave" to calcuate it.
 * Because Gridwise class and BlockWise class is also referenced in host side code, and get_warp_size is not constxpr in hostside, we can't define WaveSize, NXdlPerWave and all derived variables as static constexpr variable. we have to reference it with device function, and it causes the code change in merge_wmma_mfma1 is more than merge_wmma_mfma2.

Conclusion:
- Consider we plan to shift to CK_TILE in the future, I think it is better to support WMMA with minimum efforts, so I refer to use current solution to extend XDL kernels to RDNA architecture.
 
## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [ ] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

Snektron commented on include/ck/utility/get_id.hpp:
```diff
@@ -1,12 +1,44 @@
+#endif
+}
+
+inline __host__ index_t get_warp_size()
+{
+#if !(defined(__HIPCC_RTC__) || defined(CK_CODE_GEN_RTC))
+    hipDeviceProp_t props{};
+    int device;
+    auto status = hipGetDevice(&device);
+    if(status == hipSuccess)
+    {
+        status = hipGetDeviceProperties(&props, device);
+        if(status == hipSuccess)
+        {
+            return props.major > 9 ? 32 : 64;
+        }
+    }
+#endif
+    return 64;
+}
```
FYI I previously found out that calling hipGetDeviceProperties is pretty expensive and can cause a lot of slowdown. I suggest to replace it with `hipDeviceGetAttribute` and/or cache the the device's warp size. Also, this should probably accept a `hipStream_t` in case the current device is not the one that's associated with the stream that the kernel should launch on. rocPRIM implementation: https://github.com/ROCm/rocm-libraries/blob/b10a470ec0c83944051e8ed53e632c360ee02e3b/projects/rocprim/rocprim/include/rocprim/device/config_types.hpp#L353-L396

linqun committed:
Extend XDL kernel to Support RDNA3/4 - Part 1

Problem:
CK contains a lot of well-designed XDL kernel, it is limited on MI cards, CK also have some WMMA kernel for RDNA3/RDNA4, but the supporting is much less XDL kernel.
Consider RDNA3/RDNA4 are more and more used in ML, and our future HW architecture will close to RDNA too, we need improve the support of WMMA kernel.
After compare the WMMA kernel with XDL kernel, it looks the architecture and algorithm is almost same between them. The major differences are:
- XDL kernel based MfmaSelector + Mfma instruction, which doesn't support any WMMA instructions
- XDL kernel assume warp size is constant and it fixed to 64.
Template parameter BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave and NXdlPerWave causes warp size is hard code to 64
WaveSize is widely used as static constexpr variable, and hardcode to 64.
Proposal:
To reduce the efforts of code maintain, It is better to extend exist XDL kernels to support both GFX9 and RNDAs, instead of duplicate XDL kernel to WMMA kernel.

To eliminate the difference between GFX9 and RDNA, This PR
- Extend MfmaSelector + Mfma instruction to support all WMMA instruction.
- Ignore NXdlPerWave in device class template, and calculate it with constexpr according to BlockSize, MPerBlock, NPerBlock, MPerXDL, NPerXDL, MXdlPerWave according to current GPU target.
- To keep interface compatible, we  continue to keep NXdlPerWave as template parameters.
ps: In CK kernel, BlockSize is also used in threadgroup copy parameter, ignore BlockSize requires more code changes.

illsilin commented:
I think that's how we should have integrated support for Navis in the first place. Instead of generating new pipelines and drawing the Berlin Wall between Instincts and Navis based on MFMA vs WMMA and WarpSize 64 vs WarpSize 32, we should have used the same pipelines and just add dynamic switches between the two and add guardrails to select between instances for a given architecture. 
This will also resolve the headache with the separate sets of tests/examples that we had to maintain for different architectures.

illsilin reviewed (approved):
looks GREAT to me.
